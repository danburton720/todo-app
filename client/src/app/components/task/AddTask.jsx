import React from 'react';

import {Plus} from 'react-feather';

export default class AddTask extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            add: false,
            task: {
                changed: false,
                value: ''
            }
        };

        this.onSubmit = this.onSubmit.bind(this);
        this.onChangeTask = this.onChangeTask.bind(this);
        this.toggleAdd = this.toggleAdd.bind(this);
    }

    onSubmit(e) {
        e.preventDefault();
        if (!this.invalidTask()) {
            const data = {
                description: this.state.task.value
            };
            this.props.onSubmit(data);
            this.setState({
                add: false,
                task: {
                    changed: false,
                    value: ''
                }
            });
        }
    }

    onChangeTask(e) {
        this.setState({
            task: {
                changed: true,
                value: e.target.value
            }
        })
    }

    toggleAdd() {
        this.setState({
            add: !this.state.add
        });
    }

    invalidTask() {
        return !this.state.task.changed || this.state.task.value.length === 0;
    }

    render() {
        if (this.state.add) {
            return (
                <form onSubmit={this.onSubmit}>
                    <input
                        placeholder="Jot down a task..."
                        onChange={this.onChangeTask}
                        value={this.state.task.value}
                    />
                    <div>
                        <button disabled={this.invalidTask()} type="submit">Add</button>
                        <button onClick={this.toggleAdd} type="button">Cancel</button>
                    </div>
                </form>
            );
        }
        return (
            <div onClick={this.toggleAdd}>
                <Plus size={18}/>
                <span>Add Task</span>
            </div>
        );
    }
}