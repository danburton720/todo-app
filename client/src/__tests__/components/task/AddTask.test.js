import React from 'react';
import {mount} from 'enzyme';
import AddTask from '../../../app/components/task/AddTask';

describe('AddTask', () => {
    let mountWrapper;
    let props;
    const wrapper = () => {
        if (!mountWrapper) {
            mountWrapper = mount(<AddTask {...props}/>);
        }
        return mountWrapper;
    };

    beforeEach(() => {
        props = {
            onSubmit: jest.fn()
        };
        mountWrapper = undefined;
    });

    describe('render', () => {
        it('should return a Plus component and an \'Add Task\' span', () => {
            expect(wrapper()).toMatchSnapshot();
        });
        it('should return an add task form', () => {
            wrapper().find('div').props().onClick();
            wrapper().update();
            expect(wrapper()).toMatchSnapshot();
        });
    });

    describe('functionality', () => {
        describe('form', () => {
            beforeEach(() => {
                wrapper().find('div').props().onClick();
                wrapper().update();
            });
            describe('onSubmit', () => {
                it('should call the prop onSubmit after setting the input value to \'hello world\' ' +
                    'then remove the form and render the Plus component', () => {
                    const onChangeEvent = {
                        target: {
                            value: 'hello world'
                        }
                    };
                    const submitEvent = {
                        preventDefault: jest.fn()
                    };
                    expect(props.onSubmit).not.toHaveBeenCalled();
                    wrapper().find('input').props().onChange(onChangeEvent);
                    wrapper().update();
                    expect(wrapper().find('input').props().value).toBe('hello world')

                    wrapper().find('form').props().onSubmit(submitEvent);
                    wrapper().update();
                    expect(submitEvent.preventDefault).toHaveBeenCalledTimes(1);
                    expect(props.onSubmit).toHaveBeenCalledTimes(1);

                    expect(wrapper()).toMatchSnapshot();
                });
            });
        });
    });
});