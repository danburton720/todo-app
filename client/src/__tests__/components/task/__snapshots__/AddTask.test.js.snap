// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTask functionality form onSubmit should call the prop onSubmit after setting the input value to 'hello world' then remove the form and render the Plus component 1`] = `
<AddTask
  onSubmit={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "description": "hello world",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="addTask"
    onClick={[Function]}
  >
    <Plus
      size={18}
    >
      <svg
        fill="none"
        height={18}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="12"
          x2="12"
          y1="5"
          y2="19"
        />
        <line
          x1="5"
          x2="19"
          y1="12"
          y2="12"
        />
      </svg>
    </Plus>
    <span>
      Add Task
    </span>
  </div>
</AddTask>
`;

exports[`AddTask render should return a Plus component and an 'Add Task' span 1`] = `
<AddTask
  onSubmit={[MockFunction]}
>
  <div
    className="addTask"
    onClick={[Function]}
  >
    <Plus
      size={18}
    >
      <svg
        fill="none"
        height={18}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="12"
          x2="12"
          y1="5"
          y2="19"
        />
        <line
          x1="5"
          x2="19"
          y1="12"
          y2="12"
        />
      </svg>
    </Plus>
    <span>
      Add Task
    </span>
  </div>
</AddTask>
`;

exports[`AddTask render should return an add task form 1`] = `
<AddTask
  onSubmit={[MockFunction]}
>
  <form
    className="addTaskForm"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Jot down a task..."
      value=""
    />
    <div>
      <button
        className="primary mr1"
        disabled={true}
        type="submit"
      >
        Add
      </button>
      <button
        className="primary inverted"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</AddTask>
`;
